lms/services/application_instance_getter.py

```diff
@@ -50,75 +46,66 @@ def developer_secret(self, consumer_key):
 
-        :raise ConsumerKeyError: if the consumer key isn't in the database
+        :raise ConsumerKeyError: if the request's consumer key isn't in the DB
 
         :return: the matching LMS URL
         :rtype: str
         """
-        return self._get(consumer_key).lms_url
+        return self._get().lms_url
 
-    def provisioning_enabled(self, consumer_key):
+    def provisioning_enabled(self):
         """
-        Return ``True`` if provisioning is enabled for the given consumer key.
-
-        Return ``True`` if the provisioning feature is enabled for the given
-        consumer key, ``False`` otherwise.
+        Return ``True`` if provisioning is enabled for the current request.
 
-        :arg consumer_key: the consumer key to search for
-        :type consumer_key: str
+        Return ``True`` if the provisioning feature is enabled for the current
+        request, ``False`` otherwise.
         """
         try:
-            provisioning = self._get(consumer_key).provisioning
+            provisioning = self._get().provisioning
         except ConsumerKeyError:
             provisioning = False
         return provisioning
 
-    def shared_secret(self, consumer_key):
+    def shared_secret(self):
         """
-        Return the LTI/OAuth 1 shared secret for the given LTI consumer_key.
+        Return the LTI/OAuth 1 shared secret for the current request.
 
         This is called the ``oauth_consumer_secret`` in the OAuth 1.0 spec.
 
-        :arg consumer_key: the consumer key to search for
-        :type consumer_key: str
+        :raise ConsumerKeyError: if the request's consumer key isn't in the DB
 
-        :raise ConsumerKeyError: if the consumer key isn't in the database
-
-        :return: the matching shared secret
+        :return: the request's shared secret
         :rtype: str
         """
-        return self._get(consumer_key).shared_secret
+        return self._get().shared_secret
 
-    def _get(self, consumer_key):
+    @lru_cache(maxsize=1)
+    def _get(self):
         """
         Return the ApplicationInstance with the given consumer_key or ``None``.
 
-        :arg consumer_key: the consumer key to search for
-        :type consumer_key: str
-
         :raise ConsumerKeyError: if the consumer key isn't in the database
 
         :return: the matching ApplicationInstance or ``None``
         :rtype: :cls:`lms.models.ApplicationInstance` or ``None``
         """
-        try:
-            return (
-                self._db.query(ApplicationInstance)
-                .filter(ApplicationInstance.consumer_key == consumer_key)
-                .one()
-            )
-        except NoResultFound as err:
-            raise ConsumerKeyError() from err
+        application_instance = models.ApplicationInstance.get(
+            self._db, self._consumer_key
+        )
+
+        if application_instance is None:
+            raise ConsumerKeyError()
+
+        return application_instance
```

Comment:
Yeah that seems fine.

If `ConsumerKeyError` is moved from a services error to a models error, then it might be a layering violation for code-above-the-services (i.e. views and resources) to be catching a models error, rather than the services encapsulating the models so that code-above-services only catches services errors.

On the other hand code-above-services is also above models, so maybe there is not layering violation in code-above-services using the models (as long as it's just simple things like catching a models exception or calling a models method, not something like doing a complex SQLAlchemy query).

The only above-services thing that uses `ConsumerKeyError` anyway is `JSConfig` in `resources`.

Reply:
Done

Reply:
I decided not to do this. It turns out that `services.ConsumerKeyError` is a subset of `services.LTILaunchVerificationError` and that fact is made use of to change behaviour in a number of places.

If  `services.ConsumerKeyError(LTILaunchVerificationError)` becomes `models.ConsumerKeyError` and therefore no longer a `services.LTILaunchVerificationError` subclass, stuff breaks.

You'll end up having to keep catch `models.ConsumerKeyError` and raise `services.ConsumerKeyError(LTILaunchVerificationError)`, which removes the simplification and de-duplication benefit, and you'd have to keep a `services.ConsumerKeyError(LTILaunchVerificationError)` class around so now you have two exception classes.

Basically `ConsumerKeyError` is semantically absolutely a services-layer error and that's important